src/dataflow/components/dataflow-program.tsx

```diff
@@ -450,22 +453,34 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
 
   private hasValidOutputNodes = () => {
     const { ui } = this.stores;
-    if (!this.getNodeCount("Relay") && !this.getNodeCount("Data Storage")) {
+    if (!this.hasRelay() && !this.hasDataStorage()) {
       ui.alert("Program must contain a Relay or Data Storage node before it can be run.", "No Program Output");
       return false;
-    } else if (!this.getNodeCount("Relay") &&
+    } else if (!this.hasRelay() &&
                 this.programEditor.nodes.filter(n => (n.name === "Data Storage" && n.inputs.size <= 1)).length) {
       ui.alert("Data Storage node needs data inputs before program can be run.", "Missing Data Storage Inputs");
       return false;
     }
     return true;
   }
 
-  private runProgram = () => {
+  private prepareToRunProgram = () => {
+    const dialogPrompt = this.hasDataStorage()
+                          ? "Save dataset as"
+                          : "Save program as";
+    const baseTitle = this.getDatasetName() || this.context?.title || "program";
+    const programTitle = `${baseTitle}_${getLocalTimeStamp(Date.now())}`;
```

Comment:
We generate and present a file name to the user with a time stamp created from `Date.now()` here in `prepareToRunProgram`, but we overwrite that name and create a second file name with a call to `Date.now()` in `generateProgramData` (line 540).  This is confusing.  We tell the user that we're going to use a specific program name, but then regenerate and use a slightly different program name in a later function.

Reply:
🤦‍♂ There are two different timestamps because at the time we present the dialog we don't know the actual time of starting the program. Having gone to the trouble of asking the user for the name, however, you're right that we should use the correct one for the name of the document.